### PR TITLE
lib: smf: Add tracing support

### PR DESCRIPTION
**Pull request by wysman**
Add user-defined callback function for tracing state execution

Signed-off-by: William MARTIN <william.martin@power-lan.com>

### CODE REVIEW

**Comment by keith-zephyr** [comment]
Can you extend the existing [Tracing feature](https://docs.zephyrproject.org/latest/services/tracing/index.html) instead?

@nashif @teburd - As Tracing maintainers, is it appropriate to extend tracing to the SMF lirbary?

**Comment by wysman** [comment]
@keith-zephyr, i am not sure the tracing name is correctly chosen !

FYI: We use this code without JTAG adapter, to track the state of multiple FSM over a serial port.
This allow our end user to update a custom dashboard on PC by parsing the UART output.

**Comment by keith-zephyr** [comment]
> @keith-zephyr, i am not sure the tracing name is correctly chosen !
> 
> FYI: We use this code without JTAG adapter, to track the state of multiple FSM over a serial port. This allow our end user to update a custom dashboard on PC by parsing the UART output.

The tracing subsystem supports different backends, including a [UART backend](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/tracing/tracing_backend_uart.c).

**Comment by carlescufi** [comment]
@keith-zephyr @sambhurst please review

File: include/zephyr/smf.h [hunk 1]
```diff
@@ -107,6 +139,11 @@ struct smf_ctx {
 	 * used to track state machine context
 	 */
 	uint32_t internal;
+
+#ifdef CONFIG_SMF_TRACING_SUPPORT
+	state_tracing trace;
+	const char *name;
```
Comment (sambhurst) [inline include/zephyr/smf.h:145]: Could you add comments? Also, how is name initialized and used?

File: include/zephyr/smf.h [hunk 2]
```diff
@@ -87,8 +89,38 @@ struct smf_state {
...
+	SMF_ENTRY,
+	SMF_RUN,
+	SMF_EXIT,
+	SMF_PARENT_ENTRY,
+	SMF_PARENT_RUN,
+	SMF_PARENT_EXIT,
+	SMF_TERMINATED,
+};
+
+extern const char *smf_action_names[];
```
Comment (sambhurst) [inline include/zephyr/smf.h:113]: Add comment and how is smf_action_names initialized and used?

File: include/zephyr/smf.h [hunk 3]
```diff
@@ -87,8 +89,38 @@ struct smf_state {
 	 *	that parent's exit and entry functions do not execute.
 	 */
 	const struct smf_state *parent;
+
+#ifdef CONFIG_SMF_TRACING_SUPPORT
+	const char *name;
```
Comment (sambhurst) [inline include/zephyr/smf.h:94]: Comment and how is name initialized and used?

**Comment by wysman** [comment]
@keith-zephyr Can we close / refuse this PR as it is. 
I will propose a new one based on the tracing feature later !

**Comment by keith-zephyr** [comment]
> @keith-zephyr Can we close / refuse this PR as it is. I will propose a new one based on the tracing feature later !

Great to hear you're going to make this work with the existing tracing subsystem.  Yes, it is appropriate to close this PR while you work on a new implementation.

**Comment by wysman** [comment]
Ok

**Pull request closed without merging**